File: src/pub_sub_pkg/src/pub_sub.cpp
#include<ros/ros.h>
#include<std_msgs/Int32.h>
#include<std_msgs/Float32.h>

class s_class
	{
	ros::Subscriber sub_obj;
	ros::Publisher pub_obj;
	std_msgs::Int32 o_obj;
	std_msgs::Float32 i_obj;
//	ros::NodeHandle nh;
	public:
	s_class(int argc,char **argv)
		{
		ros::init(argc,argv,"pub_sub_class_node");
		}
	void callback(const std_msgs::Float32::ConstPtr& msg)
		{
		i_obj.data=msg->data;
		o_obj.data=2*(msg->data);
		ROS_INFO("Data subscribed= %f",msg->data);
		pub_obj.publish(o_obj);
		ROS_INFO("Data doubled after sub= %u",o_obj.data);
		}
	void pub_sub_func()
		{
		ros::NodeHandle nh;
		pub_obj=nh.advertise<std_msgs::Int32>("out_topic",5);
		sub_obj=nh.subscribe("in_topic",5,&s_class::callback,this);
		ros::spin();
		}
	};
int main(int argc,char **argv)
	{
	s_class obj(argc,argv);
	obj.pub_sub_func();
//	ros::init(argc,argv,"subscriber_class_node");
//	ros::NodeHandle nh;
//	ros::Subscriber sub=nh.subscribe("test_topic",5,&s_class::callback,&obj);
//	ros::spin();
	}

/* We can do it by two ways, either with subscribing in main, who's syntax is in comments
also, in that case we cannot define NodeHandle and subscriber object in private of class
Hence, The un-commented code is of that case along with using constructor
Apparently the initialization done in private happens before constructor, so can't include init inside constructor */ 

File: src/pub_sub_pkg/src/pub_sub.py
#! /usr/bin/env python
import rospy
from std_msgs.msg import Int32
from std_msgs.msg import Float32

class pub_sub(object):
	def __init__(self):
		self.Iobj=Int32()
		self.Fobj=Float32()
		self.Iobj.data=0
		self.Fobj.data=0
		rospy.init_node('pub_sub_node')
		self.pub_obj=rospy.Publisher('result_topic',Int32,queue_size=1)
		rospy.Subscriber('info_topic',Float32,self.callback)
#		rospy.spin()
		
	def callback(self,msg):	
		self.Fobj.data=msg.data
		rospy.loginfo("Subscribed data from info_topic = %f",self.Fobj.data)
#		self.Iobj.data=2*self.Fobj.data
#		self.pub_obj.publish(self.Iobj)
#		rospy.loginfo("Double of subscribed data is published =%u",self.Iobj.data)
#		if self.Fobj.data<20:
#			self.Iobj.data=-1
#			self.pub_obj.publish(self.Iobj)
#			rospy.loginfo("data < 20, so published to result_topic=%u",self.Iobj.data)
#		elif self.Fobj.data>20 and self.Fobj.data<30:
#			self.Iobj.data=0
#			self.pub_obj.publish(self.Iobj)
#			rospy.loginfo("data between 20&30, so published to result_topic=%u",self.Iobj.data)
#		elif self.Fobj.data>30:
#			self.Iobj.data=1
#			self.pub_obj.publish(self.Iobj)
#			rospy.loginfo("data > 30, so published to result_topic=%u",self.Iobj.data)
	def spin(self):
		rate=rospy.Rate(1)
		while not rospy.is_shutdown():
			self.Iobj.data=self.Fobj.data*2
			self.pub_obj.publish(self.Iobj)
			rospy.loginfo("Double of subscribed data is published =%u",self.Iobj.data)
			rate.sleep()
		
if __name__=='__main__':
#	try:
		obj2=pub_sub()
		obj2.spin()
		
#	except rospy.ROSInteruptException:
#		pass
